alphagov/paas-cf
# [#155630096] Fix rotation jobs

## What

I re-enabled the certificate and credentials rotation jobs.

Found the following issues:

* The smoke tests were passed the admin_user/admin_password variables but the config expects the user/password params instead. This meant the smoke tests were using the "admin" user and if the continuous smoke tests started during the rotation the failed login attempts eventually locked the admin user. (This was fixed in https://github.com/alphagov/paas-cf/pull/1271)
* The temporary test user is being deleted with the "admin" user but during a rotation this clean up can fail. Fortunately CF allows to delete a user with its own creds, so we'll self-destruct these test users.
* The UAA JWT signing keys were set with `default` and `previous` key ids in the configuration, but every issued JWT token contains the key id and can only be verified with that key. When we rotated the keys we updated the `default` key which caused all issued keys to be invalid and the `previous` key was never used (as no JWT token was ever issued with the `previous` key name). Instead of using `default` and `previous` we'll generate the SHA1 hash of these keys and use it as key id.

[1] https://github.com/cloudfoundry/uaa/blob/master/docs/UAA-APIs.rst#get-the-token-signing-key-get-token-key

## How to review

1. Run create-cloudfoundry from this branch:
    ```BRANCH=fix_rotation_jobs_155630096 make dev pipelines```

1. Run the credentials/rotate-cloudfoundry-credentials job.
   * You should be able to log in with the CF cli (scripts/cf_admin_login_dev.sh) after it

1. The continuous-smoke-tests job will fail until the cf-deploy finishes - this is expected.

1. Run the rotate-cf-internal-certs job and run the following script while running cf-deploy:
    ```while true; do date; cf curl '/v2/organizations/<admin org guid>/spaces?order-by=name&inline-relations-depth=1' | jq ".total_results"; sleep 1; done```

1. Run the delete-old-cf-internal-certs and run the same test

## Who can review

Not me.
